### PR TITLE
feat: return application for amendments (FPLA-1436)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] PR's title includes JIRA ticket number and follow [Conventional Commits](https://www.conventionalcommits.org/) specification
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-1436

### Change description ###

- Adds 'Returned' state
- Applications can be sent to returned state via 'Return application' event. Available in Submitted and Gatekeeping.
- c110a document migrated to return details object visible in the 'overview' tab.

Screenshot 2020-05-20 at 16 19 30

Screenshot 2020-05-20 at 16 19 39

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json [hunk 1]
```diff
@@ -166,5 +166,19 @@
...
+    "CaseEventID": "returnApplication",
+    "UserRole": "caseworker-publiclaw-courtadmin",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "returnApplication",
+    "UserRole": "[LASOLICITOR]",
+    "CRUD": "R"
```
Comment (Jamesnelson117) [inline ccd-definition/AuthorisationCaseEvent/AuthorisationCaseEvent-nonprod.json:182]: Left in nonprod as this story is awaiting a story to be started / complete to introduce notifiactions.

File: ccd-definition/CaseEvent/StateChange.json [hunk 2]
```diff
@@ -46,6 +46,24 @@
     "SecurityClassification": "Public",
     "EndButtonLabel": "Save and continue"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "returnApplication",
+    "Name": "Return application",
+    "Description": "Return an application to the LA",
+    "DisplayOrder": 3,
```
Comment (Jamesnelson117) [inline ccd-definition/CaseEvent/StateChange.json:55]: Display order needs corrected.

File: e2e/tests/localAuthorityCorrectsReturnedApplication_test.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+const config = require('../config.js');
```
Comment (Jamesnelson117) [inline e2e/tests/localAuthorityCorrectsReturnedApplication_test.js:1]: @douglasrice please review

File: ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json [hunk 4]
```diff
@@ -0,0 +1,50 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "ReturnApplication",
+    "CaseEventID": "returnApplication",
+    "CaseFieldID": "returnApplication",
+    "ListElementCode": "reason",
```
Comment (toireasam) [inline ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json:7]: are these names a bit generic, do we want to go with returnReason or what is your thoughts? I'm thinking outside the context of this json file it might be a bit unclear..

File: ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    "FieldDisplayOrder": 1,
+    "DisplayContext": "MANDATORY"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "ReturnApplication",
```
Comment (toireasam) [inline ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json:13]: are these names a bit generic, do we want to go with returnNote or what is your thoughts? I'm thinking outside the context of this json file it might be a bit unclear..

File: ccd-definition/State.json [hunk 6]
```diff
@@ -38,5 +38,13 @@
...
     "Description": "Deleted case state - all data is removed",
     "DisplayOrder": 5
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "RETURNED",
+    "Name": "Returned",
+    "Description": "Returned case state - LA to make changes to previously submitted application",
+    "DisplayOrder": 5
```
Comment (toireasam) [inline ccd-definition/State.json:48]: update to 6 :)

File: ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json (hunk 4 above)
Comment (Jamesnelson117) [reply ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json:7]: No I think they are fine. I'd prefer 

ReturnApplication.getReason() over ReturnApplication.getReturnReason(); 

I think the later is unnecessarily verbose.

File: e2e/pages/events/enterApplicantEvent.page.js [hunk 7]
```diff
@@ -55,4 +55,9 @@ module.exports = {
     I.fillField(this.fields(elementIndex).solicitor.dx, solicitor.dx);
     I.fillField(this.fields(elementIndex).solicitor.reference, solicitor.reference);
   },
+
+  enterPbaNumber(pbaNumber = 'PBA1234567') {
```
Comment (toireasam) [inline e2e/pages/events/enterApplicantEvent.page.js:59]: is this change needed for your story? or am i missing something :)

Comment (toireasam) [reply e2e/pages/events/enterApplicantEvent.page.js:59]: nevermind, just seen further down it's use :)

File: e2e/tests/localAuthorityCorrectsReturnedApplication_test.js (hunk 3 above)
Comment (douglasrice) [reply e2e/tests/localAuthorityCorrectsReturnedApplication_test.js:1]: e2e covers the AC scenarios for this ticket

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReturnApplicationAboutToSubmitTest.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.hmcts.reform.fpl.controllers.ReturnApplicationController.RETURN_APPLICATION;
+import static uk.gov.hmcts.reform.fpl.enums.ReturnedApplicationReasons.INCOMPLETE;
+import static uk.gov.hmcts.reform.fpl.enums.State.OPEN;
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.formatLocalDateToString;
+
+@ActiveProfiles("integration-test")
+@WebMvcTest(NoticeOfProceedingsController.class)
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReturnApplicationAboutToSubmitTest.java:26]: controller here should be return application :)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ReturnApplicationController.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+            .returnedDate(formatLocalDateToString(now(), "dd MMM YYYY"))
+            .submittedDate(formatLocalDateToString(caseData.getDateSubmitted(), "dd MMM YYYY"))
+            .build();
+    }
+
+    private String buildReturnedFileName(String fileName) {
+        String documentName = fileName.substring(0, fileName.lastIndexOf('.'));
+        String extension = fileName.substring(fileName.lastIndexOf('.'));
+
+        return documentName + "_returned" + extension;
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ReturnApplicationController.java:72]: Should these two methods be in a service of their own? that way we can test them? don't mind if you want to leave here / feel they should stay.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/ReturnedApplicationReasons.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+public enum ReturnedApplicationReasons {
+    INCORRECT("Application Incorrect"),
+    INCOMPLETE("Application Incomplete"),
+    CLARIFICATION_NEEDED("Clarification Needed");
+
+    private final String label;
+
+    ReturnedApplicationReasons(String label) {
+        this.label = label;
+    }
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/ReturnedApplicationReasons.java:12]: think you can remove this constructor if you do 

@Getter
@RequiredArgsConstructor

File: ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json (hunk 4 above)
Comment (toireasam) [reply ccd-definition/CaseEventToComplexTypes/returnApplication/ReturnApplication.json:7]: cool makes sense. :)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ReturnApplicationController.java (hunk 9 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ReturnApplicationController.java]: Yeah I think this is a good idea. And in the meantime we can also de-couple these two methods.
